Added more performance metrics with the existing query

samiura opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

Added more performance metrics with the existing query. These metrics were already fetched with the existing query but were not registered as metrics. 

- sqlserver.database.backup_or_restore.rate
- sqlserver.replica.sent.rate
- sqlserver.replica.received.rate
- sqlserver.database.execution.errors
- sqlserver.table.count
- sqlserver.page.free_list_stalls.rate
- sqlserver.database.tempdb.space
- sqlserver.database.full_scan.rate
- sqlserver.index.search.rate
- sqlserver.login.rate
- sqlserver.logout.rate
- sqlserver.database.deadlock.rate
- sqlserver.database.mirror_write_transaction.rate
- sqlserver.memory.grants_pending.count
- sqlserver.page.lookup.rate
- sqlserver.transaction.delay
- sqlserver.memory.usage
- sqlserver.database.tempdb.version_store.size
- sqlserver.lock.timeout.rate




<img width="1503" alt="Screenshot 2025-02-24 at 10 28 34" src="https://github.com/user-attachments/assets/62c7936f-ee70-4dd5-b8e6-ae1929892205" />



<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes

<!--Describe what testing was performed and which tests were added.-->
#### Testing

Tested against Micrsoft SQL Server Standard 2022

<!--Describe the documentation added.-->
#### Documentation
Updated the documentation.
<!--Please delete paragraphs that you did not use before submitting.-->

Commit:
Added more performance metrics with the existing query

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

Update .chloggen/added-more-sqlserver-receiver-preformace-metrics.yaml

Co-authored-by: Curtis Robert <crobert@splunk.com>

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

Addressing previous PR comments

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

Addressing previous PR comments

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

fixed unit tests

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

minor name changes

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

addressing previous PR comments

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

addressing previous PR comments

Signed-off-by: Samiur Arif <samiur.arif@sumologic.com>

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +255,139 @@ metrics:
+  sqlserver.full_scans:
+    enabled: false
+    description: Total full scans per second.
+    unit: “KB”
```
Great catch.. Not sure how I have made this mistake.

StefanKurek commented on .chloggen/added-more-sqlserver-receiver-preformace-metrics.yaml:
```diff
@@ -0,0 +1,27 @@
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Can you list the metrics being added here?

samiura replied:
Definitely.

crobert-1 commented on receiver/sqlserverreceiver/metadata.yaml:
I think the naming and format of the newly introduced metrics should be changed pretty considerably to match [the spec](https://opentelemetry.io/docs/specs/semconv/general/naming/).

I think the first thing we want to do is to break metrics down into namespaces, so related metrics can be found more easily.

We also want to have one metric for multi-directional, or multi-state information. View `sqlserver.database.count` as an example, it's one metric with a datapoint for each possible state.

The spec has quite a few rules around metric naming and organization that are not being met here, so it would be helpful to take a look and address as many as possible.

Overall we just want to make sure these metrics are as easy to find and understand as possible.

samiura replied:
@crobert-1 Yes! I was not happy with the naming either and will work on it along with the valuable issues addressed below by Stefan.

samiura replied:
I have pushed a new commit and tried to follow the guidelines regarding naming style but open to suggestion.

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +255,139 @@ metrics:
       input_type: string
     attributes: [database.status]
     extended_documentation: This metric is only available when the receiver is configured to directly connect to SQL Server.
+  sqlserver.active.temp.tables:
```
@samiura It seems like this would be a good candidate to have some attributes. Counting "active" and "temporary" tables implies that we could also count "inactive" or "non-temporary" tables. 

To better match existing metrics I'd suggest something like `sqlserver.table.count` with an attribute for something like `table_type` (it would probably be ok to have a single value of "temporary" and this could be expanded later) and one like `table_state` (also would probably be fine to have a single value of "active" with any other state added in the future).

Thoughts?

samiura replied:
I cannot agree more with you. Thank you for the idea.

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +255,139 @@ metrics:
+  sqlserver.active.temp.tables:
+    enabled: false
+    description: The number of active temporary tables.
+    unit: “{temp_tables}”
```
unit is probably fine as `"{tables}"`

samiura replied:
yeap +1

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +255,139 @@ metrics:
+    enabled: false
+    description: The number of active temporary tables.
+    unit: “{temp_tables}”
+    gauge:
```
Less important to me but it would probably be more accurate to label this as a non-monotonic, cumulative, sum. Although there are other similar existing metrics that are currently labeled as gauges...

samiura replied:
I can change in this issue for the newly added ones to non-monotonic.

StefanKurek reviewed (commented):
I plan on looking through all of the names soon.

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+      monotonic: false
+      value_type: int
+    attributes: []
+  sqlserver.memory.free_list_stalls.rate:
```
Actually I noticed there is another metric that starts with `sqlserver.page.buffer_cache...` so maybe that's what you should use?

samiura replied:
I would incline with `page` instead of buffer as it is more suited in this case. This metric is heavily involved with paging than buffers. ( More information [here](https://www.heelpbook.net/2015/sql-server-memory-performance-metrics-and-bottlenecks/)

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+  sqlserver.database.lock_timeout.rate:
+    enabled: false
+    description: Total number of lock timeouts per second.
+    unit: “{lock_timeouts}/s”
```
{timeouts}/s

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    gauge:
+      value_type: double
+    attributes: []
+  sqlserver.database.login.rate:
```
I don't think "Logins/sec" or "Logouts/sec" are database specific. So I would remove the `.database` from the names for them

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    gauge:
+      value_type: double
+    attributes: []
+  sqlserver.database.deadlock.rate:
```
`sqlserver.deadlock.rate`?

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    gauge:
+      value_type: double
+    attributes: []
+  sqlserver.database.mirror_write_transaction.rate:
```
`sqlserver.transaction.mirror_write.rate`?

samiura replied:
I was thinking about that but got indecisive with the definition of the SQL server manual. At any rate, this sounds better.

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+  sqlserver.database.mirror_write_transaction.rate:
+    enabled: false
+    description: Total number of mirror write transactions per second.
+    unit: “{mirror_write_transactions}/s”
```
`{transactions/s}`

samiura replied:
changed

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    gauge:
+      value_type: double
+    attributes: []
+  sqlserver.memory.grants_pending.count:
```
`sqlserver.memory.grants.pending.count`?

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    enabled: false
+    description: Total free space in temporary DB.
+    unit: “KB”
+    gauge:
```
this could probably be a non-monotonic cumulative sum.

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    enabled: false
+    description: Total number of memory grants pending.
+    unit: “{memory_grants_pending}”
+    sum:
```
Do you know what exactly this metric defines? Is it only showing current pending grants? Is it showing pending grants over the last x seconds? If it's the former, then I think how non-monotonic cumulative sum is fine. If it's the latter, then I think a gauge makes more sense.

samiura replied:
This metric states " Memory Grants Pending displays the total number of SQL Server processes that are waiting to be granted workspace in the memory."  And you can grab the value using the following SQL query.
```
SELECT object_name, counter_name, cntr_value
FROM sys.dm_os_performance_counters
WHERE [object_name] LIKE '%Memory Manager%'
AND [counter_name] = 'Memory Grants Pending'
```

To answer your question it is the former nor the latter and counter is needed. For your reference, further details can found succinctly [here](https://learn.microsoft.com/en-us/troubleshoot/sql/database-engine/performance/troubleshoot-memory-grant-issues)

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+  sqlserver.memory.grants_pending.count:
+    enabled: false
+    description: Total number of memory grants pending.
+    unit: “{memory_grants_pending}”
```
`{grants}`

samiura replied:
correct

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    enabled: false
+    description: Time consumed in transaction delays.
+    unit: ms
+    gauge:
```
Just double checking that this isn't a cumulative sum?

samiura replied:
yes i think so,

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    enabled: false
+    description: Total memory in use.
+    unit: “KB”
+    gauge:
```
Could probably be considered a non-monotonic cumulative sum. Non blocking issue though

samiura replied:
Yes, changed!

StefanKurek commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    gauge:
+      value_type: double
+    attributes: []
+  sqlserver.database.version_store_size:
```
`sqlserver.database.version_store.size` or maybe even `sqlserver.database.tempdb.version_store.size`?

samiura replied:
Changed it

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+  sqlserver.replica.sent.rate:
+    enabled: false
+    description: Bytes sent per second to a replica.
+    unit: “{bytes}/s”
```
changed

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    gauge:
+      value_type: double
+    attributes: []
+  sqlserver.database.execution_errors.count:
```
changed.

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    gauge:
+      value_type: int
+    attributes: []
+  sqlserver.replica.sent.rate:
```
I have incorporate the changes you addressed here. It makes sense to have one metric with attributes.

samiura commented:
@StefanKurek I have addressed 50% of the comments. I planned to address rest with another commit soon. Stay tuned! And also thank you so much for you diligently going through this.

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    unit: “{errors}”
+    sum:
+      aggregation_temporality: cumulative
+      monotonic: false
```
Just like the memory grants it is the former and meanings errors in the last x seconds or so.

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    gauge:
+      value_type: int
+    attributes: []
+  sqlserver.database.free_space_tempdb:
```
I agree here! I have found similar example in other receiver with attributes stating -> used/free states.

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -255,6 +263,145 @@ metrics:
+    description: The number of backups/restores per second.
+    unit: “{backups_or_restores}/s”
+    gauge:
+      value_type: int
```
definitely, i do not see any reason to disagree here.

samiura commented:
> @StefanKurek I have addressed 50% of the comments. I planned to address rest with another commit soon. Stay tuned! And also thank you so much for you diligently going through this.

@StefanKurek  I would love to have your review now as I believe I have addressed all of your comments with the latest commit.

StefanKurek reviewed (changes requested):
Getting close. A few more tweaks to the naming I think and you should be good!

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -58,6 +58,23 @@ attributes:
+    description: The status of the tempdb space usage.
+    type: string
+    enum: [free, used]
+  replica.direction:
```
:), You know what when I was adding it I was asking myself does "received, send" <-> "read, write" equivalence? And finally I thought not but since you have brought this i would rather incline using the existing attributes than adding a new attribute instead.

StefanKurek replied:
@samiura actually that was an oversight by me scanning this too quickly. I think it is fair to have a transmit/receive attribute compared to the current read/write one.  Sorry about that!

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -267,6 +284,144 @@ metrics:
+    gauge:
+      value_type: double
+    attributes: []
+  sqlserver.memory.grant.pending.count:
```
done

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -267,6 +284,144 @@ metrics:
+      monotonic: false
+      value_type: int
+    attributes: []
+  sqlserver.page.free_list_stall.rate:
```
I think you are right on this one. Thanks for providing the reference link. For some reason I was fixing this up with "sql query" memory space (temporary)  that SQL server uses.

samiura commented on receiver/sqlserverreceiver/metadata.yaml:
```diff
@@ -267,6 +284,144 @@ metrics:
+    attributes: []
+  sqlserver.replica.data.rate:
+    enabled: false
+    description: Total number of replica bytes.
```
done

StefanKurek reviewed (commented):
I think just one more change (which is actually me going back on a previous comment as seen here https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37883#discussion_r1974232177 and then I'm happy with it. Thanks for the work!

StefanKurek reviewed (approved):
I think this looks good! Thanks for all the work.

atoulme reviewed (approved):
LGTM, codeowner approved.

angristan commented on receiver/sqlserverreceiver/documentation.md:
```diff
@@ -289,6 +313,100 @@ This metric is only available when the receiver is configured to directly connec
+
+| Unit | Metric Type | Value Type | Aggregation Temporality | Monotonic |
+| ---- | ----------- | ---------- | ----------------------- | --------- |
+| “KB” | Sum | Int | Cumulative | false |
```
👋 I don't know if it's intended but there is a few occurrences of `“` being used instead of the usual `"` in this PR
